ntoskrnl/ke/amd64/traphandler.c

```diff
@@ -166,25 +166,29 @@ KiSystemCallHandler(
     }
 
     /* Get descriptor table */
-    DescriptorTable = (PVOID)((ULONG_PTR)Thread->ServiceTable + Offset);
+    DescriptorTable = &((PKSERVICE_TABLE_DESCRIPTOR)Thread->ServiceTable)[TableIndex];
 
     /* Validate the system call number */
     if (ServiceNumber >= DescriptorTable->Limit)
     {
-        /* Check if this is a GUI call */
-        if (!(Offset & SERVICE_TABLE_TEST))
+        /* Check if this is a GUI call and this is not a GUI thread yet */
+        if ((TableIndex == WIN32K_SERVICE_INDEX) &&
+            (Thread->ServiceTable == KeServiceDescriptorTable))
         {
-            /* Fail the call */
-            TrapFrame->Rax = STATUS_INVALID_SYSTEM_SERVICE;
-            return (PVOID)NtSyscallFailure;
+            /* Convert this thread to a GUI thread.
+               It is invalid to change the stack in the middle of a C function,
+               therefore we return KiConvertToGuiThread to the system call entry
+               point, which then calls the asm function KiConvertToGuiThread,
+               which allocates a new stack, switches to it, calls
+               PsConvertToGuiThread and resumes in the middle of
+               KiSystemCallEntry64 to restart the system call handling.
+               If converting fails, the system call returns a failure code. */
+            return (PVOID)KiConvertToGuiThread;
```

Comment:
I have one question (not fully related to this PR) about the x64 ASM implementation of `KiConvertToGuiThread`:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ke/amd64/trap.S;hb=f92972efcabafe1552360f8750883db3de28df9a#l1038

It creates a larger kernel stack etc. but as far as I can see, the platform-independent `PsConvertToGuiThread` already does that:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ps/win32.c;hb=f92972efcabafe1552360f8750883db3de28df9a#l32
so isn't this stack enlargement in the ASM function redundant?

In addition, after the call to `PsConvertToGuiThread`, there's a check for error,
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ke/amd64/trap.S;hb=f92972efcabafe1552360f8750883db3de28df9a#l1070
where we take the `KiConvertToGuiThreadFailed` branch. However, this branch overwrites the returned error code from `PsConvertToGuiThread` to something unrelated (`STATUS_NO_MEMORY`).

Both of these points look strange.

Reply:
hm yes, and interestingly here
https://jpassing.com/2008/10/25/how-gui-thread-conversion-on-svr03-damages-the-seh-chain/
it shows how stack-switching in the middle can introduce other subtle bugs.